subsys/net/lib/lwm2m/lwm2m_rw_senml_json.c

```diff
@@ -0,0 +1,1488 @@
+	struct lwm2m_engine_res_inst *res_inst = NULL;
+	uint8_t created;
+	int ret = 0;
+
+	/* handle resource value */
+	/* reset values */
+	created = 0U;
+
+	/* if valid, use the return value as level */
+
+	ret = lwm2m_get_or_create_engine_obj(msg, &obj_inst, &created);
+	if (ret < 0) {
+		return ret;
+	}
+
+	obj_field = lwm2m_get_engine_obj_field(obj_inst->obj, msg->path.res_id);
+	/*
+	 * if obj_field is not found,
+	 * treat as an optional resource
+	 */
+	if (!obj_field) {
+		return -ENOENT;
+	}
+
+	if (!LWM2M_HAS_PERM(obj_field, LWM2M_PERM_W)) {
+		return -EPERM;
+	}
+
+	if (!obj_inst->resources || obj_inst->resource_count == 0U) {
+		return -EINVAL;
+	}
+
+	for (int index = 0; index < obj_inst->resource_count; index++) {
+		if (obj_inst->resources[index].res_id == msg->path.res_id) {
+			res = &obj_inst->resources[index];
+			break;
+		}
+	}
+
+	if (!res) {
+		return -ENOENT;
+	}
+
+	for (int index = 0; index < res->res_inst_count; index++) {
+		if (res->res_instances[index].res_inst_id == msg->path.res_inst_id) {
+			res_inst = &res->res_instances[index];
+			break;
+		}
+	}
+
+	if (!res_inst) {
+		return -ENOENT;
+	}
+
+	/* Write the resource value */
+	ret = lwm2m_write_handler(obj_inst, res, res_inst, obj_field, msg);
+	return ret;
+}
+
+int do_write_op_senml_json(struct lwm2m_message *msg)
+{
+	struct json_in_formatter_data fd;
+	int ret = 0;
+	uint8_t name[MAX_RESOURCE_LEN + 1];
+	uint8_t base_name[MAX_RESOURCE_LEN + 1];
+	uint8_t full_name[MAX_RESOURCE_LEN + 1];
+	struct lwm2m_obj_path resource_path;
+	bool path_valid = false;
+	bool data_value = false;
+
+	struct lwm2m_block_context *block_ctx = msg->in.block_ctx;
+
+	(void)memset(&fd, 0, sizeof(fd));
+	engine_set_in_user_data(&msg->in, &fd);
+
+	if (block_ctx && block_ctx->json_flags) {
+		/* Re-load Base name and Name data from context block */
+		if (block_ctx->base_name_stored) {
+			int base_name_length = 0;
+
+			if (block_ctx->base_name_path.level == 0) {
+				base_name_length = snprintk(base_name, sizeof(base_name), "/");
+			} else if (block_ctx->base_name_path.level == 1U) {
+				base_name_length = snprintk(base_name, sizeof(base_name), "/%u/",
+							    block_ctx->base_name_path.obj_id);
+			} else if (block_ctx->base_name_path.level == 2U) {
+				base_name_length = snprintk(base_name, sizeof(base_name), "/%u/%u/",
+							    block_ctx->base_name_path.obj_id,
+							    block_ctx->base_name_path.obj_inst_id);
+			} else if (block_ctx->base_name_path.level == 3U) {
+				base_name_length =
+					snprintk(base_name, sizeof(base_name), "/%u/%u/%u",
+						 block_ctx->base_name_path.obj_id,
+						 block_ctx->base_name_path.obj_inst_id,
+						 block_ctx->base_name_path.res_id);
+			} else {
+				base_name_length =
+					snprintk(base_name, sizeof(base_name), "/%u/%u/%u/%u",
+						 block_ctx->base_name_path.obj_id,
+						 block_ctx->base_name_path.obj_inst_id,
+						 block_ctx->base_name_path.res_id,
+						 block_ctx->base_name_path.res_inst_id);
+			}
```

Comment:
Maybe it'd make sense to extract this as a utility function, and re-use at line 1319?

Reply:
yes good point I will do that. Only change is that path level need to be given by own parameter. Good spot